File: ESTRUCURAS_DE_DATOS/LISTAS_DOBLES_ENLAZADAS/typeScript/cliente.ts
/**
 * Nodo que representa un cliente en la cola circular.
 */
class Cliente {
  nombre: string;
  anterior: Cliente | null;
  siguiente: Cliente | null;

  constructor(nombre: string) {
    this.nombre = nombre;
    this.anterior = null;
    this.siguiente = null;
  }
}

/**
 * Lista circular doble que maneja turnos de clientes.
 */
class ColaCircular {
  inicio: Cliente | null = null;

  /**
   * Agrega un cliente al final de la cola.
   */
  agregarCliente(nombre: string): void {
    const nuevo = new Cliente(nombre);

    if (this.inicio === null) {
      this.inicio = nuevo;
      nuevo.siguiente = nuevo;
      nuevo.anterior = nuevo;
    } else {
      const ultimo = this.inicio.anterior as Cliente;
      ultimo.siguiente = nuevo;
      nuevo.anterior = ultimo;
      nuevo.siguiente = this.inicio;
      this.inicio.anterior = nuevo;
    }

    console.log(`✅ Cliente "${nombre}" agregado a la cola.`);
  }

  /**
   * Atiende al siguiente cliente (mueve al final de la cola).
   */
  atenderCliente(): void {
    if (this.inicio === null) {
      console.log("⚠️ No hay clientes en la cola.");
      return;
    }

    console.log(`👩‍💼 Atendiendo a: ${this.inicio.nombre}`);
    this.inicio = this.inicio.siguiente;
  }

  /**
   * Muestra todos los clientes en espera.
   */
  mostrarCola(): void {
    if (this.inicio === null) {
      console.log("La cola está vacía.");
      return;
    }

    let actual = this.inicio;
    const lista: string[] = [];
    do {
      lista.push(actual.nombre);
      actual = actual.siguiente as Cliente;
    } while (actual !== this.inicio);

    console.log("🧾 Cola de clientes:", lista.join(" -> "));
  }
}

// --- PRUEBA ---
const cola = new ColaCircular();
cola.agregarCliente("Carlos");
cola.agregarCliente("Ana");
cola.agregarCliente("Luis");

cola.mostrarCola();
cola.atenderCliente();
cola.mostrarCola();

File: ESTRUCURAS_DE_DATOS/LISTAS_DOBLES_ENLAZADAS/PYTHON/cliente.py
# ==============================================================
# 🧾 LISTA DOBLE CIRCULAR - SISTEMA DE TURNOS
# ==============================================================

class Cliente:
    """Nodo que representa un cliente en la cola circular."""
    def __init__(self, nombre):
        self.nombre = nombre
        self.anterior = None
        self.siguiente = None


class ColaCircular:
    """Maneja una cola circular de clientes."""
    def __init__(self):
        self.inicio = None

    def agregar_cliente(self, nombre):
        """Agrega un cliente al final de la cola."""
        nuevo = Cliente(nombre)
        if self.inicio is None:
            self.inicio = nuevo
            nuevo.siguiente = nuevo
            nuevo.anterior = nuevo
        else:
            ultimo = self.inicio.anterior
            ultimo.siguiente = nuevo
            nuevo.anterior = ultimo
            nuevo.siguiente = self.inicio
            self.inicio.anterior = nuevo
        print(f"✅ Cliente '{nombre}' agregado a la cola.")

    def atender_cliente(self):
        """Atiende al siguiente cliente (mueve al final)."""
        if self.inicio is None:
            print("⚠️ No hay clientes en la cola.")
            return
        print(f"👩‍💼 Atendiendo a: {self.inicio.nombre}")
        self.inicio = self.inicio.siguiente

    def mostrar_cola(self):
        """Muestra todos los clientes en la cola circular."""
        if self.inicio is None:
            print("📭 La cola está vacía.")
            return

        actual = self.inicio
        lista = []
        while True:
            lista.append(actual.nombre)
            actual = actual.siguiente
            if actual == self.inicio:
                break
        print("🧾 Cola de clientes:", " -> ".join(lista))


# ---------- CÓDIGO DE PRUEBA ----------
if __name__ == "__main__":
    cola = ColaCircular()
    cola.agregar_cliente("Carlos")
    cola.agregar_cliente("Ana")
    cola.agregar_cliente("Luis")

    cola.mostrar_cola()
    cola.atender_cliente()
    cola.mostrar_cola()
